diff -Naur freetype-2.6.5.orig/include/freetype/config/ftoption.h freetype-2.6.5/include/freetype/config/ftoption.h
--- freetype-2.6.5.orig/include/freetype/config/ftoption.h	2016-09-06 21:20:54.818152155 -0400
+++ freetype-2.6.5/include/freetype/config/ftoption.h	2016-09-06 21:25:20.214133258 -0400
@@ -92,7 +92,7 @@
   /* This is done to allow FreeType clients to run unmodified, forcing     */
   /* them to display normal gray-level anti-aliased glyphs.                */
   /*                                                                       */
-/* #define FT_CONFIG_OPTION_SUBPIXEL_RENDERING */
+#define FT_CONFIG_OPTION_SUBPIXEL_RENDERING
 
 
   /*************************************************************************/
@@ -631,7 +631,7 @@
   /* [1] http://www.microsoft.com/typography/cleartype/truetypecleartype.aspx */
   /*                                                                       */
 /* #define TT_CONFIG_OPTION_SUBPIXEL_HINTING  1         */
-/* #define TT_CONFIG_OPTION_SUBPIXEL_HINTING  2         */
+#define TT_CONFIG_OPTION_SUBPIXEL_HINTING  2
 /* #define TT_CONFIG_OPTION_SUBPIXEL_HINTING  ( 1 | 2 ) */
 
 
