diff -Naur before/debian/rules after/debian/rules
--- before/debian/rules	2015-04-19 21:23:38.855936164 +0000
+++ after/debian/rules	2015-04-19 22:00:35.705961983 +0000
@@ -386,6 +386,7 @@
 	icc_loading_with_symlink.diff \
 	icedtea-4953367.patch \
 	icedtea-override-redirect-compiz.patch \
+	add-fontconfig-support.diff \
 	jdk-freetypeScaler-crash.diff \
 	libpcsclite-dlopen.diff \
 	nonreparenting-wm.diff \
